src/compiler/array.rs

```diff
@@ -574,3 +617,376 @@ fn copy(
+        );
+        // []
+
+        ptr
+    };
+
+    let second_target_data_ptr = {
+        let ptr = compiler
+            .memory
+            .allocate_symbol(Type::PrimitiveType(PrimitiveType::UInt32));
+
+        compiler
+            .memory
+            .read(compiler.instructions, data_ptr(&new_arr).memory_addr, 1);
+        // [data_ptr]
+        compiler.memory.read(
+            compiler.instructions,
+            start.memory_addr,
+            start.type_.miden_width(),
+        );
+        // [start]
+        compiler.instructions.push(Instruction::U32CheckedAdd);
+        // [data_ptr + start]
+        compiler.memory.write(
+            compiler.instructions,
+            ptr.memory_addr,
+            &[ValueSource::Stack],
+        );
+        // []
+
+        ptr
+    };
+
+    let second_length = {
+        let len = compiler
+            .memory
+            .allocate_symbol(Type::PrimitiveType(PrimitiveType::UInt32));
+
+        compiler.memory.read(
+            compiler.instructions,
+            length(&arr).memory_addr,
+            length(&arr).type_.miden_width(),
+        );
+        // [length]
+        compiler.memory.read(
+            compiler.instructions,
+            delete_count.memory_addr,
+            delete_count.type_.miden_width(),
+        );
+        // [delete_count, length]
+        compiler.instructions.push(Instruction::U32CheckedSub);
+        // [length - delete_count]
+        compiler.memory.write(
+            compiler.instructions,
+            len.memory_addr,
+            &[ValueSource::Stack],
+        );
+        // []
+
+        len
+    };
+
+    copy(
+        compiler,
+        &second_source_data_ptr,
+        &second_length,
+        &second_target_data_ptr,
+        &capacity(&new_arr),
+        element_type.miden_width(),
+    )?;
+
+    compiler.memory.write(
+        compiler.instructions,
+        length(&new_arr).memory_addr,
+        &vec![ValueSource::Memory(new_arr_len.memory_addr)],
+    );
+
+    compiler.memory.read(
+        compiler.instructions,
+        new_arr.memory_addr,
+        new_arr.type_.miden_width(),
+    );
+    compiler.memory.write(
+        compiler.instructions,
+        arr.memory_addr,
+        &vec![ValueSource::Stack; arr.type_.miden_width() as usize],
+    );
+
+    return Ok(array_of_deletions);
+}
+
+pub(crate) fn slice(
+    compiler: &mut Compiler,
+    arr: &Symbol,
+    start: Option<Symbol>,
+    end: Option<&Symbol>,
+) -> Result<Symbol> {
+    ensure_eq_type!(arr, Type::Array(_));
+    let start = start.unwrap_or(uint32::new(compiler, 0));
+    ensure_eq_type!(start, Type::PrimitiveType(PrimitiveType::UInt32));
+    if let Some(end) = end {
+        ensure_eq_type!(end, Type::PrimitiveType(PrimitiveType::UInt32));
+    }
+
+    let element_type = element_type(&arr.type_);
+
+    let actual_start = {
+        let actual_start = compiler
+            .memory
+            .allocate_symbol(Type::PrimitiveType(PrimitiveType::UInt32));
```

Comment:
Are this symbol and `actual_end` really needed? Seems that everything could be done in stack, no?